### PR TITLE
Adding template for TFS multi-server deployment

### PR DESCRIPTION
**Pull request by chrisamert**
Creates a TFS deployment with three VMS.  One Domain Controller, one running SQL Server Standard, and one with TFS

### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### CODE REVIEW

File: tfs-dual-server/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+  },
+  "variables": {
+    "vnetName": "tfsStandardVNet",
+    "loadBalancerName": "tfsStandardLoadBalancer",
+    "subnetName": "tfsStandardSubnet",
+    "loadBalancerBackendName": "LoadBalancerBackend",
+    "availabilitySetName": "tfsStandardAvailabilitySet",
+    "dcIpAddress": "10.0.0.4",
+    "nestedTemplates": {
+      "networkTemplate": "[concat(parameters('_artifactsLocation'), '/nestedtemplates/virtualNetwork.json', parameters('_artifactsLocationSasToken'))]",
```
Comment (bmoore-msft) [inline tfs-dual-server/azuredeploy.json]: use the uri() to build the urls

File: tfs-dual-server/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+      "domainControllerTemplate": "[concat(parameters('_artifactsLocation'), '/nestedtemplates/domainController.json', parameters('_artifactsLocationSasToken'))]",
+      "sqlServerTemplate": "[concat(parameters('_artifactsLocation'), '/nestedtemplates/sqlServer.json', parameters('_artifactsLocationSasToken'))]",
+      "tfsServerTemplate": "[concat(parameters('_artifactsLocation'), '/nestedtemplates/tfs.json', parameters('_artifactsLocationSasToken'))]"
+    }
+  },
+  "resources": [
+    {
+      "name": "Network",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tfs-dual-server/azuredeploy.json]: update apiVersion

File: tfs-dual-server/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+          "dnsServers": {
+            "value": []
+          }
+        }
+      }
+    },
+    {
+      "name": "DomainController",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tfs-dual-server/azuredeploy.json]: update apiVersion

File: tfs-dual-server/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+          "_artifactsLocationSasToken": {
+            "value": "[parameters('_artifactsLocationSasToken')]"
+          }
+        }
+      }
+    },
+    {
+      "name": "UpdateDNS",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tfs-dual-server/azuredeploy.json]: update apiVersion

File: tfs-dual-server/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+              "[variables('dcIpAddress')]"
+            ]
+          }
+        }
+      }
+    },
+    {
+      "name": "SQLServer",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tfs-dual-server/azuredeploy.json]: update apiVersion

File: tfs-dual-server/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+          "_artifactsLocationSasToken": {
+            "value": "[parameters('_artifactsLocationSasToken')]"
+          }
+        }
+      }
+    },
+    {
+      "name": "TfsServer",
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2015-01-01",
```
Comment (bmoore-msft) [inline tfs-dual-server/azuredeploy.json]: update apiVersion

File: tfs-dual-server/nestedtemplates/domainController.json [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+      "metadata": {
+        "description": "The sasToken required to access _artifactsLocation.  When the template is deployed using the accompanying scripts, a sasToken will be automatically generated."
+      }
+    }
+  },
+  "variables": {
+    "networkInterfaceName": "[concat(parameters('vmName'), 'nic')]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',parameters('existingVnetName'))]",
+    "loadBalancerID": "[resourceId('Microsoft.Network/loadBalancers',parameters('existingLoadBalancerName'))]",
+    "subnetRef": "[concat(variables('vnetID'),'/subnets/', parameters('existingSubnetName'))]"
```
Comment (bmoore-msft) [inline tfs-dual-server/nestedtemplates/domainController.json]: simplify, just use resourceId()

File: tfs-dual-server/nestedtemplates/virtualNetwork.json [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    }
+  },
+  "variables": {
+    "publicIpName": "publicIp",
+    "lbFrontEndName": "LoadBalancerFrontEnd",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',parameters('vnetName'))]",
+    "publicIpID": "[resourceId('Microsoft.Network/publicIPAddresses',variables('publicIpName'))]",
+    "loadBalancerID": "[resourceId('Microsoft.Network/loadBalancers',parameters('loadBalancerName'))]",
+    "frontEndIPConfigID": "[concat(variables('loadBalancerID'),'/frontendIPConfigurations/', variables('lbFrontEndName'))]",
+    "subnetRef": "[concat(variables('vnetID'), '/subnets/', parameters('subnetName'))]"
```
Comment (bmoore-msft) [inline tfs-dual-server/nestedtemplates/virtualNetwork.json]: simplify, just use resourceId()

**Comment by chrisamert** [comment]
This template is expected to take a while to complete (often > 1hr), and seems to be hitting a timeout in the CI build.  Is there any way to increase that timeout?

File: tfs-dual-server/nestedtemplates/domainController.json [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    "_artifactsLocationSasToken": {
+      "type": "securestring",
+      "metadata": {
+        "description": "The sasToken required to access _artifactsLocation.  When the template is deployed using the accompanying scripts, a sasToken will be automatically generated."
+      }
+    }
+  },
+  "variables": {
+    "networkInterfaceName": "[concat(parameters('vmName'), 'nic')]",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',parameters('existingVnetName'))]",
```
Comment (bmoore-msft) [inline tfs-dual-server/nestedtemplates/domainController.json]: remove this it's not used

File: tfs-dual-server/nestedtemplates/virtualNetwork.json [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+      "type": "array",
+      "metadata": {
+        "description": "List of DNS servers in the network"
+      }
+    }
+  },
+  "variables": {
+    "publicIpName": "publicIp",
+    "lbFrontEndName": "LoadBalancerFrontEnd",
+    "vnetID": "[resourceId('Microsoft.Network/virtualNetworks',parameters('vnetName'))]",
```
Comment (bmoore-msft) [inline tfs-dual-server/nestedtemplates/virtualNetwork.json]: remove this it's not used

**Comment by bmoore-msft** [comment]
We increased the timeout on the builds - so it does complete now...  a couple pr comments then we're good.

**Pull request merged**